### ISSUE TITLE
Allow all players to challenge

### ISSUE THREAD

**Issue by pkolano** [cdot/Xanado#22]
Currently, only the player after the one who played can challenge a word.  Can you please make it so any player can challenge?  Thanks!

**Comment by cdot** [comment]
As I read the rules of Scrabble, only the following player can challenge. I think the same is true of Lexulous and Words with Friends, but I might be wrong there.

At the moment the game can't proceed until the following player has either played or challenged, which keeps it synchronous and simple. Letting anyone challenge might be difficult, as you have to deal with the case where challenges arrive asynchronously.

**Comment by pkolano** [comment]
I'm looking online, but I'm not really seeing anything suggesting only the player after may challenge for players >= 3.  Most of the ones I see are talking about 2 player tournament games where it obviously has to be the next player.  Can you send the link you found for the rules that indicates only the next player can challenge for 3+ player games?  On the hasbro site (https://scrabble.hasbro.com/en-us/rules), it just says:

> Any play may be challenged before the next player starts a turn. If the play challenged is unacceptable, the challenged player takes back his or her tiles and loses that turn. If the play challenged is acceptable, the challenger loses his or her next turn. Consult the dictionary for challenges only. All words made in one play are challenged simultaneously. If any word is unacceptable, then the entire play is unacceptable. Only one turn is lost on any challenge. 

This seems to indicate anyone can challenge as it talks in general terms about "the challenger".  We've always played anyone can challenge.  Otherwise, the order is even more critical (and makes the game more lucky) because some people aren't afraid of losing their turn and challenge more often and some know more words (in both cases, unlucky to be before them and lucky to not be).

I understand this may be somewhat tricky, but asynchronous communication already exists in the chat box, so don't see why a challenge can't just be another similar asynchronous message that can be processed up until the next player plays.  Thanks.

**Comment by cdot** [comment]
You are right, the "next player" comes from tournament rules. I can't find anything that explicitly gives that rule for multi-player games, probably because part of the charm of Scrabble is the ability to make up "house rules"! It would be nice to extend the challenge code to support open challenge - for example, only double challenge is supported at the moment, but Dingle challenge is common in tournaments. Yes, it can be done asynchronously, it's just a bit fiddly. A PR that implements it would be welcome! ;-)

**Comment by cdot** [comment]
That required a lot of refactoring, and a load more unit tests, but it seems to be working OK.

**Comment by cdot** [comment]
All that work to get the server working, and I plain forgot to update the UI! Really should write some UI unit tests - some day!

**Comment by pkolano** [comment]
I've been told that there is still a bug with this.  During a game, someone challenged out of order (i.e. not the player after) and the challenge succeeded, but then the player after wasn't able to complete their turn.  They could put letters down, but finish turn didn't do anything so everyone had to pass around until it got back to them.  Please take a look when you get a chance.  Thanks!

**Comment by pkolano** [comment]
Just experienced this myself and after the challenge, it shows the timer running for the person who challenged so maybe it is assuming the next player is the challenging player or something.

**Comment by cdot** [comment]
Hi, I'd appreciate it if you could open a new issue for bugs, rather that annotating a closed issue. I don't visit the repository all that often (the last few weeks have been exceptional) and can't keep track otherwise. I have created a new issue for your report.
